finetune/encoding/input_encoder.py

```diff
@@ -223,23 +226,41 @@ def __getstate__(self):
         return {"Encoder": None}
 
 
+def get_relevant_context_for_chunk(context, encoded_output):
+    # this is not always right given the tokenization might change the tokens and thus the length
+    start_idx = encoded_output.char_starts[1]
+    final_idx = encoded_output.char_locs[-2]
+    new_context = []
+    for span in context:
+        if span['end'] >= start_idx:
+            new_context.append(span)
+    return new_context
+
+
 def tokenize_context(context, encoded_output, config):
     """ Tokenize the context corresponding to a single sequence of text """
+    # in the edge case where the chunk is just a single end token, we don't need to alter our context chunk
+    if len(encoded_output.char_locs) > 1:
+        context = get_relevant_context_for_chunk(context, encoded_output)
     seq_len = len(encoded_output.token_ids)
     context_keys = list(k for k in sorted(context[0].keys()) if k not in ['token', 'start', 'end'])
     context_by_char_loc = sorted([(c['end'], [c[k] for k in context_keys]) for c in context], key=lambda c: c[0])
-    # default context is the sequence majority
+    # default context is set by user in config
     default_context = [config.default_context[k] for k in context_keys]
     current_char_loc = 0
     tokenized_context = []
-    for char_loc in encoded_output.char_locs:
+    for token, char_loc in zip(encoded_output.tokens, encoded_output.char_locs):
         # Note: this assumes that the tokenization will never lump multiple tokens into one
+        # (this would not be the case if multiple context spans make up the same token)
         if char_loc == -1:
             tokenized_context.append(default_context)
+        elif token in ['\n</w>', 'Ċ', 'Ġ']:
```

Comment:
can we have tokenizers define their own of these special tokens, so it is clearer what needs to be done when a new tokenizer is added?

Reply:
good idea, will add that